### PR TITLE
add a migrator to move assets to ATP

### PR DESCRIPTION
**Pull request by birarda**

### CODE REVIEW

**Comment by hifi-gustavo** [comment]
Build console output for commit: 549043f25533ce5506c70b02db11671ac76cd843 available at: https://gist.github.com/be5896546dda56c4917b<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-6075.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-6075.dmg

File: libraries/networking/src/AssetUpload.cpp [hunk 1]
```diff
@@ -17,65 +17,89 @@
...
-                    default:
-                        _error = FileOpenError;
-                        break;
-                }
-            }
+        if (file.open(QIODevice::ReadOnly)) {
             
-            if (_error == NoError && hash == hashData(_data).toHex()) {
-                saveToCache(getUrl(hash, _extension), _data);
-            }
```
Comment (Atlante45) [inline libraries/networking/src/AssetUpload.cpp]: You deleted that block in the merge.

**Comment by hifi-gustavo** [comment]
Build console output for commit: 14503053902da26c91daa4e47206ee4cc2495f95 available at: https://gist.github.com/be5896546dda56c4917b<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-6075.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-6075.dmg

File: interface/src/assets/ATPAssetMigrator.cpp [hunk 2]
```diff
@@ -0,0 +1,263 @@
...
+static const QString ENTITIES_OBJECT_KEY = "Entities";
+static const QString MODEL_URL_KEY = "modelURL";
+static const QString MESSAGE_BOX_TITLE = "ATP Asset Migration";
+
+void ATPAssetMigrator::loadEntityServerFile() {
+    auto filename = QFileDialog::getOpenFileName(_dialogParent, "Select an entity-server content file to migrate",
+                                                 QString(), QString("Entity-Server Content (*.gz)"));
+    
+    if (!filename.isEmpty()) {
+        qDebug() << "Selected filename for ATP asset migration: " << filename;
```
Comment (Atlante45) [inline interface/src/assets/ATPAssetMigrator.cpp]: Use `qCDebug`

File: interface/src/assets/ATPAssetMigrator.cpp [hunk 3]
```diff
@@ -0,0 +1,263 @@
...
+        
+        // try to open the file at the given filename
+        QFile modelsFile { filename };
+        
+        if (modelsFile.open(QIODevice::ReadOnly)) {
+            QByteArray compressedJsonData = modelsFile.readAll();
+            QByteArray jsonData;
+            
+            if (!gunzip(compressedJsonData, jsonData)) {
+                QMessageBox::warning(_dialogParent, "Error", "The file at" + filename + "was not in gzip format.");
```
Comment (Atlante45) [inline interface/src/assets/ATPAssetMigrator.cpp]: Are all the son save zipped now?

File: interface/src/assets/ATPAssetMigrator.cpp [hunk 4]
```diff
@@ -0,0 +1,263 @@
...
+            for (auto jsonValue : _entitiesArray) {
+                QJsonObject entityObject = jsonValue.toObject();
+                QString modelURLString = entityObject.value(MODEL_URL_KEY).toString();
+                
+                if (!modelURLString.isEmpty()) {
+                    QUrl modelURL = QUrl(modelURLString);
+                    
+                    if (!_ignoredUrls.contains(modelURL)
+                        && (modelURL.scheme() == URL_SCHEME_HTTP || modelURL.scheme() == URL_SCHEME_HTTPS
+                            || modelURL.scheme() == URL_SCHEME_FILE || modelURL.scheme() == URL_SCHEME_FTP)) {
```
Comment (Atlante45) [inline interface/src/assets/ATPAssetMigrator.cpp]: Feels like you'd better check if it's not APT.

Comment (birarda) [inline interface/src/assets/ATPAssetMigrator.cpp]: I'm only migrating if it matches one of the schemes there - which checks that it isn't already ATP.

File: interface/src/assets/ATPAssetMigrator.cpp [hunk 5]
```diff
@@ -0,0 +1,267 @@
...
+                            // we already have a request out for this asset, just store the QJsonValueRef
+                            // so we can do the hash replacement when the request comes back
+                            _pendingReplacements.insert(modelURL, jsonValue);
+                        } else if (_uploadedAssets.contains(modelURL)) {
+                            // we already have a hash for this asset
+                            // so just do the replacement immediately
+                            entityObject[MODEL_URL_KEY] = _uploadedAssets.value(modelURL).toString();
+                            jsonValue = entityObject;
+                        } else if (wantsToMigrateResource(modelURL)) {
+                            auto request = ResourceManager::createResourceRequest(this, modelURL);
```
Comment (Atlante45) [inline interface/src/assets/ATPAssetMigrator.cpp]: Check request is not `nullptr`

File: interface/src/assets/ATPAssetMigrator.cpp (hunk 3 above)
Comment (birarda) [inline interface/src/assets/ATPAssetMigrator.cpp]: I'm looking into that, the migrator as written only handles gzipped JSON files (only lets you chose .json.gz when selecting file). I'm not sure for the usage this feature will get it's worth handling any other case.

**Comment by hifi-gustavo** [comment]
Build console output for commit: 76bfc6218a8e685acb5a09473aefafeb06ffda91 available at: https://gist.github.com/be5896546dda56c4917b<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-6075.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-6075.dmg

File: interface/src/assets/ATPAssetMigrator.cpp [hunk 6]
```diff
@@ -0,0 +1,267 @@
...
+
+void ATPAssetMigrator::assetUploadFinished(AssetUpload *upload, const QString& hash) {
+    if (upload->getError() == AssetUpload::NoError) {
+
+        const auto& modelURL = _originalURLs[upload];
+        
+        // successfully uploaded asset - make any required replacements found in the pending replacements
+        auto values = _pendingReplacements.values(modelURL);
+        
+        QString atpURL = QString("%1:%2.%3").arg(URL_SCHEME_ATP).arg(hash).arg(upload->getExtension());
```
Comment (Atlante45) [inline interface/src/assets/ATPAssetMigrator.cpp]: That's why I created `::getUrl(hash, ext)`

**Comment by hifi-gustavo** [comment]
Build console output for commit: cc9376707744c28511edb506902c9c3bfe3654a2 available at: https://gist.github.com/be5896546dda56c4917b<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-6075.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-6075.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: d22c60204137b7e62bfdeb87cb42c30a74fd7616 available at: https://gist.github.com/be5896546dda56c4917b<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-6075.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-6075.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 2caa7f6d647ab3c75d078caab670388e47cf755b available at: https://gist.github.com/be5896546dda56c4917b<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-6075.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-6075.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: d84994d8e0a1da6bdef40cdb13951e3281882116 available at: https://gist.github.com/be5896546dda56c4917b<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-6075.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-6075.dmg

**Pull request merged**